keps/0013-pull-based-audit-events-stream.md

```diff
@@ -0,0 +1,199 @@
+---
+kep-number: 13
+title: pull based audit events stream
+authors:
+  - "@CaoShuFeng"
+owning-sig: sig-auth
+participating-sigs:
+  - sig-api-machinery
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2018-06-11
+last-updated: 2018-06-11
+status: provisional
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+
+---
+
+# Pull based audit events stream
+
+## Table of Contents
+
+* [Pull based audit events stream](#pull-based-audit-events-stream)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Make audit backends dynamicly registerable](#make-audit-backends-dynamicly-registerable)
+    * [endpoints and query parameters](#endpoints-and-query-parameters)
+    * [Bandwidth limit Configuration](#bandwidth-limit-configuration)
+    * [User Stories](#user-stories)
+      * [Story 1](#story-1)
+      * [Story 2](#story-2)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+This proposal aims at providing a pull-based stream for audit logging. This would look similar to a watch request to the API server, and provide stream of audit events. The stream does not return any previous events (i.e. no associated storage), it would only stream events that occurred after the stream was opened. Two non resource-url endpoints are installed to provide the stream. '/audits/' for all audit events generated in the API server, and '/audits/{namespace}' for audits events in the specified namespace. And query parameters could be used to get subset of audit events.
+
+## Motivation
+The current audit backends (log & webhook) require very high privileges (file access to the master) or prior configuration (webhook). There are a number of use cases for dynamic and unprivileged access to the audit logs. Examples include:
+- allow tenants to get audit events in their own namespace
+- provide a subset of audit events, so it will be easier to query and analyze (e.g. https://github.com/kubernetes/kubernetes/issues/56683)
+
+
+### Goals
+- Provide a new pull based audit event stream, which returns events that occurred after the stream was opened.
+- Allow users go get a subset of audit events.
+- Access control. Allow unprivileged users to get audits in their own namespace.
+- Bandwidth limit in server scope or per-namespace.
+
+### Non-Goals
+- Access control at user/group level(e.g. allow user A to only get audits events from users A,B,C)
+
+
+## Proposal
```

Comment:
Can you add more implementation details? In particular, I'd like to see:

- Is this a new [audit backend](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/apiserver/plugin/pkg/audit)?
- How are events routed to the watchers?
- How are the watch connections maintained? (Does apimachinery solve this?)

Reply:
What about the aggregated API server case?
- will the kube-aggregator aggregate audit events?

Reply:
If we turn this into a resource endpoints (as @tallclair suggests above), it will be part of discovery. And in theory the aggregator could pick it up. But those api groups will conflict. So the same resource can only be used for one API server at a time. So this topic needs some more thought, to be at least extensible in the future.

Reply:
As written above, storing to disk sounds like doubling the etcd load, our bottleneck resource in big clusters.

Reply:
In order to get a first-pass of this feature in, we could simply say that the streaming endpoint only supports metadata-level, and then it can be handled entirely by kube-aggregator. This simplification also addresses the concerns about sensitive fields.

Reply:
> we could simply say that the streaming endpoint only supports metadata-level,

Test in my environment that could work. Thanks for the idea.

> In that case, we have no guarantees that we collected all events with a pull mechanism because we don't see the instance anymore in such a case.

We only trace audit events in aggregator, and request distributed to aggregated apiservers will be logged at metadata-level.

Reply:
> We only trace audit events in aggregator

Misread @tallclair suggestion. Do you have a WIP PR? Sounds reasonably easy to do, if there are no surprises.

Reply:
Hmm, I forgot about HA masters. I think this still doesn't address that. I guess this would need some pubsub/message solution - does anything else in kubernetes solve a similar problem, or does everything just rely on etcd?

Reply:
> just etcd

Reply:
Actually, this seems pretty similar to (non-audit) `Events`.